File: solutions/python/protein-translation/1/protein_translation.py
def proteins(strand):
    chunks = chunkerize(strand, 3)
    translation = []

    for chunk in chunks:
        translated_chunk = translate(chunk)
        if translated_chunk == "STOP":
            break
        translation.append(translated_chunk)

    return translation


def chunkerize(strand, size=3):
    if len(strand) == size:
        return [strand]
    else:
        chunks = []
        for i in range(0, len(strand), size):
            chunk = strand[i:i + size]
            chunks.append(chunk)

        return chunks


def translate(codon):
    match codon:
        case "AUG":
            return "Methionine"
        case "UUU" | "UUC":
            return "Phenylalanine"
        case "UUA" | "UUG":
            return "Leucine"
        case "UCU" | "UCC" | "UCA" | "UCG":
            return "Serine"
        case "UAU" | "UAC":
            return "Tyrosine"
        case "UGU" | "UGC":
            return "Cysteine"
        case "UGG":
            return "Tryptophan"
        case "UAA" | "UAG" | "UGA":
            return "STOP"


proteins("UUUUAGAUG")

File: solutions/javascript/protein-translation/1/protein-translation.js
export const translate = (sequence) => {
  if (sequence === "" || sequence === undefined) return [];

  const chunks = codonChunks(sequence);
  const proteins = assembleProteins(chunks);

  if (proteins.includes("STOP")) {
    return proteins.slice(0, proteins.indexOf("STOP"));
  }

  if (sequence.length % 3 !== 0 || proteins.includes("ERROR")) {
    throw new Error("Invalid codon");
  }
  return proteins;
};

function codonChunks(sequence) {
  const chunks = [];
  while (sequence?.length >= 3) {
    chunks.push(sequence.slice(0, 3));
    sequence = sequence.slice(3);
  }
  return chunks;
}

function assembleProteins(chunks) {
  const proteins = [];
  for (let chunk of chunks) {
    switch (chunk) {
      case "AUG":
        proteins.push("Methionine");
        break;
      case "UUU":
      case "UUC":
        proteins.push("Phenylalanine");
        break;
      case "UUA":
      case "UUG":
        proteins.push("Leucine");
        break;
      case "UCU":
      case "UCC":
      case "UCA":
      case "UCG":
        proteins.push("Serine");
        break;
      case "UAU":
      case "UAC":
        proteins.push("Tyrosine");
        break;
      case "UGU":
      case "UGC":
        proteins.push("Cysteine");
        break;
      case "UGG":
        proteins.push("Tryptophan");
        break;
      case "UAA":
      case "UAG":
      case "UGA":
        proteins.push("STOP");
        break;
      default:
        proteins.push("ERROR");
    }
  }
  return proteins;
}
